src/lsp-server.ts

```diff
@@ -38,6 +38,47 @@ import { computeCallers, computeCallees } from './calls';
 import { IServerOptions } from './utils/configuration';
 import { TypeScriptVersion, TypeScriptVersionProvider } from './utils/versionProvider';
 import { TypeScriptAutoFixProvider } from './features/fix-all';
+import { LspClient, ProgressReporter } from './lsp-client';
+
+class ServerInitializingIndicator {
+    private _loadingProjectName?: string;
+    private _progressReporter?: ProgressReporter;
+
+    constructor(private lspClient: LspClient) {}
+
+    public reset(): void {
+        if (this._loadingProjectName) {
+            this._loadingProjectName = undefined;
+            if (this._progressReporter) {
+                this._progressReporter.end();
+                this._progressReporter = undefined;
+            }
+        }
+    }
+
+    /**
+     * Signal that a project has started loading.
+     */
+    public startedLoadingProject(projectName: string | undefined): void {
+        // TS projects are loaded sequentially. Cancel existing task because it should always be resolved before
+        // the incoming project loading task is.
+        this.reset();
+
+        this._loadingProjectName = projectName;
+        this._progressReporter = this.lspClient.createProgressReporter();
+        this._progressReporter.begin('Initializing JS/TS language features');
+    }
+
+    public finishedLoadingProject(projectName: string | undefined): void {
```

Comment:
can the project type be just a string?
```suggestion
    public finishedLoadingProject(projectName: string): void {
```

Reply:
If I'm to trust Typescript types then this should be OK.